amethyst_network/src/network_socket.rs

```diff
@@ -0,0 +1,188 @@
+                        InternalSocketEvent::Stop => break 'outer,
+                    }
+                }
+
+                // receive
+                let mut buf = [0 as u8; 2048];
+                loop {
+                    match socket.recv_from(&mut buf) {
+                        // Data received
+                        Ok((amt, src)) => {
+                            receive_queue
+                                .send(RawEvent {
+                                    byte_count: amt,
+                                    data: buf[..amt].iter().cloned().collect(),
+                                    source: src,
+                                }).unwrap();
+                        }
+                        Err(e) => {
+                            if e.kind() == ErrorKind::WouldBlock {
+                                break;
+                            } else {
+                                error!("Could not receive datagram: {}", e);
+                            }
+                        }
+                    }
+                }
+            }
+        });
+
+        Ok(NetSocketSystem {
+            filters,
+            tx: tx1,
+            rx: rx2,
+        })
+    }
+}
+
+impl<'a, E> System<'a> for NetSocketSystem<E>
+where
+    E: Send + Sync + Serialize + Clone + DeserializeOwned + PartialEq + 'static,
+{
+    type SystemData = (WriteStorage<'a, NetConnection<E>>);
+
+    fn setup(&mut self, res: &mut Resources) {
+        Self::SystemData::setup(res);
+    }
+
+    fn run(&mut self, mut net_connections: Self::SystemData) {
+        for mut net_connection in (&mut net_connections).join() {
+            let target = net_connection.target.clone();
+
+            if net_connection.state == ConnectionState::Connected
+                || net_connection.state == ConnectionState::Connecting
+            {
+                self.tx
+                    .send(InternalSocketEvent::SendEvents {
+                        target,
+                        events: net_connection.send_buffer_early_read().cloned().collect(),
+                    }).unwrap();
+            } else if net_connection.state == ConnectionState::Disconnected {
+                self.tx.send(InternalSocketEvent::Stop).unwrap();
+            }
+        }
+
+        for raw_event in self.rx.try_iter() {
+            let mut matched = false;
+            // Get the NetConnection from the source
+            for mut net_connection in (&mut net_connections).join() {
+                // We found the origin
+                if net_connection.target == raw_event.source {
+                    matched = true;
+                    // Get the event
+                    let net_event = deserialize_event::<E>(raw_event.data.as_slice());
+                    match net_event {
+                        Ok(ev) => {
+                            // Filter events
+                            let mut filtered = false;
+
+                            if !filtered {
```

Comment:
There's no opportunity for `filtered` to be true here.  Why are we doing this?

Reply:
Thanks for catching this, I've removed it. I want to look back into filtering once this is merged.